### ISSUE TITLE
MongoMemoryReplSet keeps hanging indefinitely - MongoServerSelectionError: Server selection timed out after 30000 ms

### ISSUE THREAD

**Issue by hoffnung8493** [nodkz/mongodb-memory-server#460]
## Versions

- NodeJS: 14.15.1
- mongodb-memory-server-*: 6.9.6(also tried with 7.0.0-beta.26)
- mongodb: 0.0.0 // default
- mongoose: 5.11.15
- system: MacOS Big Sur 11.1(intel)

package: mongo-memory-server

## What is the Problem?

I get the following error. `MongoServerSelectionError: Server selection timed out after 30000 ms`.
According to the logs, the program is stuck at `await replSet.waitUntilRunning)`.
I have tried single instance mongo and it works fine.

## Code Example

```ts
let replSet: any
beforeAll(async () => {
  console.log('beforeAll')
  const replSet = new MongoMemoryReplSet({
    replSet: { storageEngine: 'wiredTiger' },
  })

  await replSet.waitUntilRunning() //its stuck here!!
  console.log("replica set ready")
  const uri = await replSet.getUri()
  console.log({ uri })
  await mongoose.connect(uri, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useFindAndModify: false,
    useCreateIndex: true,
  })
  console.log('mongodb connected')
})
```

## Do you know *why* it happenes?

No, I had this same setup few months ago in another laptop and it worked fined. I wonder if it has to do with macOS Big Sur

**Comment by hasezoey** [comment]
without extra debug logs it is hard to say what the problem is, please provide [debug output](https://github.com/nodkz/mongodb-memory-server#enable-debug-mode) (from latest (beta) version)

**Comment by hoffnung8493** [comment]
This feature is nice. Here are the debug logs
```
beforeAll
  MongoMS:MongoMemoryReplSet Autostarting MongoMemoryReplSet. +0ms
  MongoMS:MongoMemoryReplSet start +2ms
  MongoMS:MongoMemoryReplSet init +0ms
  MongoMS:MongoMemoryReplSet   starting server 1 of 1... +1ms
  MongoMS:MongoMemoryReplSet    instance opts: {
  auth: false,
  args: [ '--oplogSize', '1' ],
  dbName: 'a61770ff-bf39-4e9b-a4c4-830b74df482d',
  ip: '127.0.0.1',
  replSet: 'testset',
  storageEngine: 'wiredTiger'
} +0ms
  MongoMS:MongoMemoryServer Called MongoMemoryServer.start() method +0ms
  MongoMS:MongoMemoryServer starting with port 56093, since undefined was locked: 56093 +2ms
  MongoMS:MongoMemoryServer Starting MongoDB instance with following options: {"port":56093,"dbName":"a61770ff-bf39-4e9b-a4c4-830b74df482d","ip":"127.0.0.1","storageEngine":"wiredTiger","replSet":"testset","dbPath":"/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--55582-zoIvCWAA8XXu","tmpDir":{"name":"/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--55582-zoIvCWAA8XXu"},"uri":"mongodb://127.0.0.1:56093/a61770ff-bf39-4e9b-a4c4-830b74df482d?"} +1ms
  MongoMS:MongoBinary MongoBinary options: {
  "downloadDir": "/Users/sihoonkim/Documents/Rocketship/backend/node_modules/.cache/mongodb-memory-server/mongodb-binaries",
  "platform": "darwin",
  "arch": "x64",
  "version": "4.0.14",
  "checkMD5": false
} +0ms
  MongoMS:MongoBinary MongoBinary: Download lock removed +3ms
  MongoMS:MongoBinary MongoBinary: Mongod binary path: "/Users/sihoonkim/Documents/Rocketship/backend/node_modules/.cache/mongodb-memory-server/mongodb-binaries/4.0.14/mongod" +0ms
  MongoMS:MongoInstance Mongo[56093]: Called MongoInstance._launchKiller(parent: 55582, child: 55592): +0ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:22.021+0900 I CONTROL  [main] Automatically disabling TLS 1.0, to force-enable TLS 1.0 specify --sslDisabledProtocols 'none'
  MongoMS:MongoInstance  +21ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:22.034+0900 I CONTROL  [initandlisten] MongoDB starting : pid=55592 port=56093 dbpath=/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--55582-zoIvCWAA8XXu 64-bit host=gimsihun-ui-MacBookPro.local
  MongoMS:MongoInstance  +11ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:22.034+0900 I CONTROL  [initandlisten] db version v4.0.14
  MongoMS:MongoInstance 2021-03-18T18:14:22.034+0900 I CONTROL  [initandlisten] git version: 1622021384533dade8b3c89ed3ecd80e1142c132
  MongoMS:MongoInstance 2021-03-18T18:14:22.034+0900 I CONTROL  [initandlisten] allocator: system
  MongoMS:MongoInstance 2021-03-18T18:14:22.034+0900 I CONTROL  [initandlisten] modules: none
  MongoMS:MongoInstance 2021-03-18T18:14:22.034+0900 I CONTROL  [initandlisten] build environment:
  MongoMS:MongoInstance 2021-03-18T18:14:22.034+0900 I CONTROL  [initandlisten]     distarch: x86_64
  MongoMS:MongoInstance 2021-03-18T18:14:22.034+0900 I CONTROL  [initandlisten]     target_arch: x86_64
  MongoMS:MongoInstance 2021-03-18T18:14:22.034+0900 I CONTROL  [initandlisten] options: { net: { bindIp: "127.0.0.1", port: 56093 }, replication: { oplogSizeMB: 1, replSet: "testset" }, security: { authorization: "disabled" }, storage: { dbPath: "/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--55582-zoIvCWAA8XXu", engine: "wiredTiger" } }
  MongoMS:MongoInstance  +1ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:22.035+0900 I STORAGE  [initandlisten] wiredtiger_open config: create,cache_size=15872M,cache_overflow=(file_max=0M),session_max=20000,eviction=(threads_min=4,threads_max=4),config_base=false,statistics=(fast),log=(enabled=true,archive=true,path=journal,compressor=snappy),file_manager=(close_idle_time=100000),statistics_log=(wait=0),verbose=(recovery_progress),
  MongoMS:MongoInstance  +1ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:22.581+0900 I STORAGE  [initandlisten] WiredTiger message [1616058862:581906][55592:0x116033e00], txn-recover: Set global recovery timestamp: 0
  MongoMS:MongoInstance  +546ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:22.731+0900 I RECOVERY [initandlisten] WiredTiger recoveryTimestamp. Ts: Timestamp(0, 0)
  MongoMS:MongoInstance  +149ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:22.913+0900 I STORAGE  [initandlisten] createCollection: local.startup_log with generated UUID: 6e1751b2-2dd5-4e3a-abf2-a62e2fe0c50a
  MongoMS:MongoInstance  +182ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.011+0900 I FTDC     [initandlisten] Initializing full-time diagnostic data capture with directory '/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--55582-zoIvCWAA8XXu/diagnostic.data'
  MongoMS:MongoInstance  +98ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.012+0900 I STORAGE  [initandlisten] createCollection: local.replset.oplogTruncateAfterPoint with generated UUID: 1a20c016-2d1a-49b4-9201-c3b7b2904d90
  MongoMS:MongoInstance  +1ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.121+0900 I STORAGE  [initandlisten] createCollection: local.replset.minvalid with generated UUID: b095e408-1914-44a6-ac3f-421ef2fe5916
  MongoMS:MongoInstance  +110ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.222+0900 I STORAGE  [initandlisten] createCollection: local.replset.election with generated UUID: a818543a-137f-4849-a6bc-cb40cfc601a3
  MongoMS:MongoInstance  +100ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.346+0900 I REPL     [initandlisten] Did not find local initialized voted for document at startup.
  MongoMS:MongoInstance 2021-03-18T18:14:23.346+0900 I REPL     [initandlisten] Did not find local Rollback ID document at startup. Creating one.
  MongoMS:MongoInstance  +124ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.346+0900 I STORAGE  [initandlisten] createCollection: local.system.rollback.id with generated UUID: f16adfea-1147-4d14-a99d-a43efb30b637
  MongoMS:MongoInstance  +0ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.461+0900 I REPL     [initandlisten] Initialized the rollback ID to 1
  MongoMS:MongoInstance 2021-03-18T18:14:23.461+0900 I REPL     [initandlisten] Did not find local replica set configuration document at startup;  NoMatchingDocument: Did not find replica set configuration document in local.system.replset
  MongoMS:MongoInstance  +115ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.462+0900 I CONTROL  [LogicalSessionCacheRefresh] Sessions collection is not set up; waiting until next sessions refresh interval: Replication has not yet been configured
  MongoMS:MongoInstance  +1ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.462+0900 I NETWORK  [initandlisten] waiting for connections on port 56093
  MongoMS:MongoInstance 2021-03-18T18:14:23.462+0900 I CONTROL  [LogicalSessionCacheReap] Sessions collection is not set up; waiting until next sessions reap interval: config.system.sessions does not exist
  MongoMS:MongoInstance  +0ms
  MongoMS:MongoInstance Mongo[56093]: MongodbInstance: Instance is ready! +0ms
  MongoMS:MongoMemoryReplSet Initializing replica set. +1s
  MongoMS:MongoMemoryServer Called MongoMemoryServer.ensureInstance() method +1s
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.474+0900 I NETWORK  [listener] connection accepted from 127.0.0.1:56094 #1 (1 connection now open)
  MongoMS:MongoInstance  +18ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.480+0900 I NETWORK  [conn1] received client metadata from 127.0.0.1:56094 conn1: { driver: { name: "nodejs", version: "3.6.3" }, os: { type: "Darwin", name: "darwin", architecture: "x64", version: "20.2.0" }, platform: "'Node.js v14.15.1, LE (unified)" }
  MongoMS:MongoInstance  +0ms
(node:55582) UnhandledPromiseRejectionWarning: MongoServerSelectionError: Server selection timed out after 30000 ms
    at Timeout._onTimeout (/Users/sihoonkim/Documents/Rocketship/backend/node_modules/mongodb/lib/core/sdam/topology.js:438:30)
    at listOnTimeout (internal/timers.js:554:17)
    at processTimers (internal/timers.js:497:7)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:55582) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:55582) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:53.476+0900 I NETWORK  [conn1] end connection 127.0.0.1:56094 (0 connections now open)
  MongoMS:MongoInstance  +30s
```

**Comment by hasezoey** [comment]
guessing from the debug log, did you really test `7.0.0-beta.26`? because that was the version that actually should have fixed this timeout 

PS: from what version (mongodb-memory-server package) is the debug output?

**Comment by hoffnung8493** [comment]
`
beforeAll
  MongoMS:MongoMemoryReplSet Autostarting MongoMemoryReplSet. +0ms
  MongoMS:MongoMemoryReplSet start +2ms
  MongoMS:MongoMemoryReplSet init +0ms
  MongoMS:MongoMemoryReplSet   starting server 1 of 1... +1ms
  MongoMS:MongoMemoryReplSet    instance opts: {
  auth: false,
  args: [ '--oplogSize', '1' ],
  dbName: 'a61770ff-bf39-4e9b-a4c4-830b74df482d',
  ip: '127.0.0.1',
  replSet: 'testset',
  storageEngine: 'wiredTiger'
} +0ms
  MongoMS:MongoMemoryServer Called MongoMemoryServer.start() method +0ms
  MongoMS:MongoMemoryServer starting with port 56093, since undefined was locked: 56093 +2ms
  MongoMS:MongoMemoryServer Starting MongoDB instance with following options: {"port":56093,"dbName":"a61770ff-bf39-4e9b-a4c4-830b74df482d","ip":"127.0.0.1","storageEngine":"wiredTiger","replSet":"testset","dbPath":"/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--55582-zoIvCWAA8XXu","tmpDir":{"name":"/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--55582-zoIvCWAA8XXu"},"uri":"mongodb://127.0.0.1:56093/a61770ff-bf39-4e9b-a4c4-830b74df482d?"} +1ms
  MongoMS:MongoBinary MongoBinary options: {
  "downloadDir": "/Users/sihoonkim/Documents/Rocketship/backend/node_modules/.cache/mongodb-memory-server/mongodb-binaries",
  "platform": "darwin",
  "arch": "x64",
  "version": "4.0.14",
  "checkMD5": false
} +0ms
  MongoMS:MongoBinary MongoBinary: Download lock removed +3ms
  MongoMS:MongoBinary MongoBinary: Mongod binary path: "/Users/sihoonkim/Documents/Rocketship/backend/node_modules/.cache/mongodb-memory-server/mongodb-binaries/4.0.14/mongod" +0ms
  MongoMS:MongoInstance Mongo[56093]: Called MongoInstance._launchKiller(parent: 55582, child: 55592): +0ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:22.021+0900 I CONTROL  [main] Automatically disabling TLS 1.0, to force-enable TLS 1.0 specify --sslDisabledProtocols 'none'
  MongoMS:MongoInstance  +21ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:22.034+0900 I CONTROL  [initandlisten] MongoDB starting : pid=55592 port=56093 dbpath=/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--55582-zoIvCWAA8XXu 64-bit host=gimsihun-ui-MacBookPro.local
  MongoMS:MongoInstance  +11ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:22.034+0900 I CONTROL  [initandlisten] db version v4.0.14
  MongoMS:MongoInstance 2021-03-18T18:14:22.034+0900 I CONTROL  [initandlisten] git version: 1622021384533dade8b3c89ed3ecd80e1142c132
  MongoMS:MongoInstance 2021-03-18T18:14:22.034+0900 I CONTROL  [initandlisten] allocator: system
  MongoMS:MongoInstance 2021-03-18T18:14:22.034+0900 I CONTROL  [initandlisten] modules: none
  MongoMS:MongoInstance 2021-03-18T18:14:22.034+0900 I CONTROL  [initandlisten] build environment:
  MongoMS:MongoInstance 2021-03-18T18:14:22.034+0900 I CONTROL  [initandlisten]     distarch: x86_64
  MongoMS:MongoInstance 2021-03-18T18:14:22.034+0900 I CONTROL  [initandlisten]     target_arch: x86_64
  MongoMS:MongoInstance 2021-03-18T18:14:22.034+0900 I CONTROL  [initandlisten] options: { net: { bindIp: "127.0.0.1", port: 56093 }, replication: { oplogSizeMB: 1, replSet: "testset" }, security: { authorization: "disabled" }, storage: { dbPath: "/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--55582-zoIvCWAA8XXu", engine: "wiredTiger" } }
  MongoMS:MongoInstance  +1ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:22.035+0900 I STORAGE  [initandlisten] wiredtiger_open config: create,cache_size=15872M,cache_overflow=(file_max=0M),session_max=20000,eviction=(threads_min=4,threads_max=4),config_base=false,statistics=(fast),log=(enabled=true,archive=true,path=journal,compressor=snappy),file_manager=(close_idle_time=100000),statistics_log=(wait=0),verbose=(recovery_progress),
  MongoMS:MongoInstance  +1ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:22.581+0900 I STORAGE  [initandlisten] WiredTiger message [1616058862:581906][55592:0x116033e00], txn-recover: Set global recovery timestamp: 0
  MongoMS:MongoInstance  +546ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:22.731+0900 I RECOVERY [initandlisten] WiredTiger recoveryTimestamp. Ts: Timestamp(0, 0)
  MongoMS:MongoInstance  +149ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:22.913+0900 I STORAGE  [initandlisten] createCollection: local.startup_log with generated UUID: 6e1751b2-2dd5-4e3a-abf2-a62e2fe0c50a
  MongoMS:MongoInstance  +182ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.011+0900 I FTDC     [initandlisten] Initializing full-time diagnostic data capture with directory '/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--55582-zoIvCWAA8XXu/diagnostic.data'
  MongoMS:MongoInstance  +98ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.012+0900 I STORAGE  [initandlisten] createCollection: local.replset.oplogTruncateAfterPoint with generated UUID: 1a20c016-2d1a-49b4-9201-c3b7b2904d90
  MongoMS:MongoInstance  +1ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.121+0900 I STORAGE  [initandlisten] createCollection: local.replset.minvalid with generated UUID: b095e408-1914-44a6-ac3f-421ef2fe5916
  MongoMS:MongoInstance  +110ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.222+0900 I STORAGE  [initandlisten] createCollection: local.replset.election with generated UUID: a818543a-137f-4849-a6bc-cb40cfc601a3
  MongoMS:MongoInstance  +100ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.346+0900 I REPL     [initandlisten] Did not find local initialized voted for document at startup.
  MongoMS:MongoInstance 2021-03-18T18:14:23.346+0900 I REPL     [initandlisten] Did not find local Rollback ID document at startup. Creating one.
  MongoMS:MongoInstance  +124ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.346+0900 I STORAGE  [initandlisten] createCollection: local.system.rollback.id with generated UUID: f16adfea-1147-4d14-a99d-a43efb30b637
  MongoMS:MongoInstance  +0ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.461+0900 I REPL     [initandlisten] Initialized the rollback ID to 1
  MongoMS:MongoInstance 2021-03-18T18:14:23.461+0900 I REPL     [initandlisten] Did not find local replica set configuration document at startup;  NoMatchingDocument: Did not find replica set configuration document in local.system.replset
  MongoMS:MongoInstance  +115ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.462+0900 I CONTROL  [LogicalSessionCacheRefresh] Sessions collection is not set up; waiting until next sessions refresh interval: Replication has not yet been configured
  MongoMS:MongoInstance  +1ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.462+0900 I NETWORK  [initandlisten] waiting for connections on port 56093
  MongoMS:MongoInstance 2021-03-18T18:14:23.462+0900 I CONTROL  [LogicalSessionCacheReap] Sessions collection is not set up; waiting until next sessions reap interval: config.system.sessions does not exist
  MongoMS:MongoInstance  +0ms
  MongoMS:MongoInstance Mongo[56093]: MongodbInstance: Instance is ready! +0ms
  MongoMS:MongoMemoryReplSet Initializing replica set. +1s
  MongoMS:MongoMemoryServer Called MongoMemoryServer.ensureInstance() method +1s
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.474+0900 I NETWORK  [listener] connection accepted from 127.0.0.1:56094 #1 (1 connection now open)
  MongoMS:MongoInstance  +18ms
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:23.480+0900 I NETWORK  [conn1] received client metadata from 127.0.0.1:56094 conn1: { driver: { name: "nodejs", version: "3.6.3" }, os: { type: "Darwin", name: "darwin", architecture: "x64", version: "20.2.0" }, platform: "'Node.js v14.15.1, LE (unified)" }
  MongoMS:MongoInstance  +0ms
(node:55582) UnhandledPromiseRejectionWarning: MongoServerSelectionError: Server selection timed out after 30000 ms
    at Timeout._onTimeout (/Users/sihoonkim/Documents/Rocketship/backend/node_modules/mongodb/lib/core/sdam/topology.js:438:30)
    at listOnTimeout (internal/timers.js:554:17)
    at processTimers (internal/timers.js:497:7)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:55582) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:55582) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
  MongoMS:MongoInstance Mongo[56093]: STDOUT: 2021-03-18T18:14:53.476+0900 I NETWORK  [conn1] end connection 127.0.0.1:56094 (0 connections now open)
  MongoMS:MongoInstance  +30s
`

**Comment by hoffnung8493** [comment]
Sorry I missed the logs for the beta version. Here are the logs:
```
beforeAll
  MongoMS:MongoMemoryReplSet waitUntilRunning: stopped +0ms
Error: State is not "running" or "init" - cannot wait on something that dosnt start
    at MongoMemoryReplSet.<anonymous> (/Users/sihoonkim/Documents/Rocketship/backend/node_modules/mongodb-memory-server-core/src/MongoMemoryReplSet.ts:465:15)
    at Generator.next (<anonymous>)
    at fulfilled (/Users/sihoonkim/Documents/Rocketship/backend/node_modules/tslib/tslib.js:111:62)
```

**Comment by hoffnung8493** [comment]
Sorry I missed the logs for the beta version. Here are the logs:
```
beforeAll
  MongoMS:MongoMemoryReplSet waitUntilRunning: stopped +0ms
Error: State is not "running" or "init" - cannot wait on something that dosnt start
    at MongoMemoryReplSet.<anonymous> (/Users/sihoonkim/Documents/Rocketship/backend/node_modules/mongodb-memory-server-core/src/MongoMemoryReplSet.ts:465:15)
    at Generator.next (<anonymous>)
    at fulfilled (/Users/sihoonkim/Documents/Rocketship/backend/node_modules/tslib/tslib.js:111:62)
```

**Comment by hasezoey** [comment]
the beta error is completely different that the other, in the beta (or from 7.0.0 onward), you have to manually call `.start`, or directly with `.create` to start an instance / replset

**Comment by hoffnung8493** [comment]
Ok thanks, let me try that. But there is no fix for the latest stable version? It would be great if this is documented. I think most developers would want to test with replica set due to multi document transaction requiring a replica set. And also I believe that `.create` is not defined in the typescript type/interface. I will try again with `ts-ignore` and leave the results here

**Comment by hasezoey** [comment]
`.create` is in typescript, because this project is written in typescript & export with full types, but `.create` is an new function in 7.0.0 - before it was an auto-start on `new Class()` which could not be disabled

**Comment by hoffnung8493** [comment]
ok I tried `await replSet.start()` and I got the following logs this time.
```
beforeAll
  MongoMS:MongoMemoryReplSet start +0ms
  MongoMS:MongoMemoryReplSet initAllServers +2ms
  MongoMS:MongoMemoryReplSet   starting server 1 of 1 +0ms
  MongoMS:MongoMemoryReplSet    instance opts: {
  auth: false,
  args: [],
  dbName: '888462fe-65eb-46fa-af1d-7c31dfd2f3bc',
  ip: '127.0.0.1',
  replSet: 'testset',
  storageEngine: 'wiredTiger'
} +0ms
  MongoMS:MongoMemoryReplSet initAllServers: waiting for all servers to finish starting +2ms
  MongoMS:MongoMemoryServer start: Called .start() method +0ms
  MongoMS:MongoMemoryServer _startUpInstance: Called MongoMemoryServer._startUpInstance() method +0ms
  MongoMS:MongoMemoryServer getStartOptions +1ms
  MongoMS:MongoMemoryServer _startUpInstance: Creating new MongoDB instance with options: {"instance":{"dbPath":"/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--59874-ejK88nw0MeZP","ip":"127.0.0.1","port":56498,"storageEngine":"wiredTiger","replSet":"testset","args":[],"auth":false},"binary":{},"spawn":{}} +2ms
  MongoMS:MongoInstance Mongo[56498]: run +0ms
  MongoMS:MongoBinary getPath +0ms
  MongoMS:DryMongoBinary generateDownloadPath: Generating Download Path, preferGlobal: "true" +0ms
  MongoMS:DryMongoBinary generateDownloadPath: Paths: {
  legacyHomeCache: '',
  modulesCache: '/Users/sihoonkim/Documents/Rocketship/backend/node_modules/.cache/mongodb-memory-server/4.0.20/mongod',
  relative: '/Users/sihoonkim/Documents/Rocketship/backend/mongodb-binaries/4.0.20/mongod',
  resolveConfig: ''
} +1ms
  MongoMS:DryMongoBinary generateDownloadPath: no existing binary for version "4.0.20" was found +1ms
  MongoMS:DryMongoBinary generateDownloadPath: using global (preferGlobal) "" +0ms
  MongoMS:MongoBinary getPath: MongoBinary options: {
  "version": "4.0.20",
  "downloadDir": "/Users/sihoonkim/Documents",
  "os": {
    "os": "darwin"
  },
  "arch": "x64",
  "systemBinary": "",
  "platform": "darwin",
  "checkMD5": false
} +3ms
  MongoMS:DryMongoBinary locateBinary: Trying to locate Binary for version "4.0.20" +0ms
  MongoMS:DryMongoBinary locateBinary: running generateDownloadPath +0ms
  MongoMS:DryMongoBinary generateDownloadPath: Generating Download Path, preferGlobal: "true" +0ms
  MongoMS:DryMongoBinary generatePaths: resolveConfigValue is not empty +1ms
  MongoMS:DryMongoBinary generateDownloadPath: Paths: {
  legacyHomeCache: '',
  modulesCache: '/Users/sihoonkim/Documents/Rocketship/backend/node_modules/.cache/mongodb-memory-server/4.0.20/mongod',
  relative: '/Users/sihoonkim/Documents/Rocketship/backend/mongodb-binaries/4.0.20/mongod',
  resolveConfig: '/Users/sihoonkim/Documents/4.0.20/mongod'
} +0ms
  MongoMS:DryMongoBinary generateDownloadPath: no existing binary for version "4.0.20" was found +0ms
  MongoMS:DryMongoBinary generateDownloadPath: using resolveConfig (DOWNLOAD_DIR) "/Users/sihoonkim/Documents/4.0.20/mongod" +1ms
  MongoMS:DryMongoBinary locateBinary: could not find an existing binary +0ms
  MongoMS:MongoBinary getPath: "RUNTIME_DOWNLOAD" is "true", trying to download +2ms
  MongoMS:MongoBinary download +0ms
  MongoMS:MongoBinary download: Waiting to acquire Download lock for file "/Users/sihoonkim/Documents/4.0.20.lock" +0ms
  MongoMS:LockFile lock: Locking file "/Users/sihoonkim/Documents/4.0.20.lock" +0ms
  MongoMS:LockFile checkLock: for file "/Users/sihoonkim/Documents/4.0.20.lock" with uuid: "undefined" +0ms
  MongoMS:LockFile createLock: trying to create a lock file for "/Users/sihoonkim/Documents/4.0.20.lock" +0ms
  MongoMS:LockFile createLock: Lock File Created for file "/Users/sihoonkim/Documents/4.0.20.lock" +2ms
  MongoMS:MongoBinary download: Download lock acquired +3ms
  MongoMS:MongoBinary download: Adding version 4.0.20 to cache +0ms
  MongoMS:MongoBinaryDownload getMongodPath +0ms
  MongoMS:MongoBinaryDownload startDownload +0ms
  MongoMS:MongoBinaryDownloadUrl Using "mongodb-osx-ssl-x86_64-4.0.20.tgz" as the Archive String +0ms
  MongoMS:MongoBinaryDownloadUrl Using "https://fastdl.mongodb.org" as the mirror +0ms
  MongoMS:MongoBinaryDownload download +1ms
  MongoMS:MongoBinaryDownload download: Downloading: "https://fastdl.mongodb.org/osx/mongodb-osx-ssl-x86_64-4.0.20.tgz" +1ms
  MongoMS:MongoBinaryDownload httpDownload +0ms
  MongoMS:MongoBinaryDownload httpDownload: trying to download https://fastdl.mongodb.org/osx/mongodb-osx-ssl-x86_64-4.0.20.tgz +0ms
  MongoMS:MongoBinaryDownload httpDownload: moved "/Users/sihoonkim/Documents/mongodb-osx-ssl-x86_64-4.0.20.tgz.downloading" to "/Users/sihoonkim/Documents/mongodb-osx-ssl-x86_64-4.0.20.tgz" +3s
  MongoMS:MongoBinaryDownload makeMD5check: Checking MD5 of downloaded binary... +0ms
  MongoMS:MongoBinaryDownload makeMD5check: checkMD5 is disabled +0ms
  MongoMS:MongoBinaryDownload extract +1ms
  MongoMS:MongoBinaryDownload extract: archive: "/Users/sihoonkim/Documents/mongodb-osx-ssl-x86_64-4.0.20.tgz" final: "/Users/sihoonkim/Documents/4.0.20/mongod" +0ms
  MongoMS:MongoBinaryDownload extractTarGz +0ms
  MongoMS:MongoBinary download: Removing Download lock +4s
  MongoMS:LockFile unlock: Unlocking file "/Users/sihoonkim/Documents/4.0.20.lock" +4s
  MongoMS:LockFile checkLock: for file "/Users/sihoonkim/Documents/4.0.20.lock" with uuid: "5f3b27d0-fbba-4612-9cc1-4df58ca57fcb" +0ms
  MongoMS:LockFile checkLock: Lock File Already exists, and is for *this* process, with uuid: "5f3b27d0-fbba-4612-9cc1-4df58ca57fcb" +1ms
  MongoMS:LockFile unlock: Lock Status was "availableInstance" for file "/Users/sihoonkim/Documents/4.0.20.lock" +0ms
  MongoMS:LockFile unlockCleanup: for file "/Users/sihoonkim/Documents/4.0.20.lock" +0ms
  MongoMS:MongoBinary download: Download lock removed +3ms
  MongoMS:MongoBinary getPath: Mongod binary path: "/Users/sihoonkim/Documents/4.0.20/mongod" +0ms
  MongoMS:MongoInstance Mongo[56498]: run: Starting Processes +4s
  MongoMS:MongoInstance Mongo[56498]: _launchMongod: Launching Mongod Process +0ms
  MongoMS:MongoInstance Mongo[56498]: prepareCommandArgs +0ms
  MongoMS:MongoInstance Mongo[56498]: prepareCommandArgs: final arugment array:["--port","56498","--dbpath","/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--59874-ejK88nw0MeZP","--storageEngine","wiredTiger","--bind_ip","127.0.0.1","--noauth","--replSet","testset"] +0ms
  MongoMS:MongoInstance Mongo[56498]: _launchKiller: Launching Killer Process (parent: 59874, child: 59899) +11ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:55.404+0900 I CONTROL  [main] Automatically disabling TLS 1.0, to force-enable TLS 1.0 specify --sslDisabledProtocols 'none'"" +611ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:55.414+0900 I CONTROL  [initandlisten] MongoDB starting : pid=59899 port=56498 dbpath=/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--59874-ejK88nw0MeZP 64-bit host=gimsihun-ui-MacBookPro.local
  MongoMS:MongoInstance 2021-03-18T18:54:55.414+0900 I CONTROL  [initandlisten] db version v4.0.20
  MongoMS:MongoInstance 2021-03-18T18:54:55.414+0900 I CONTROL  [initandlisten] git version: e2416422da84a0b63cde2397d60b521758b56d1b
  MongoMS:MongoInstance 2021-03-18T18:54:55.414+0900 I CONTROL  [initandlisten] allocator: system
  MongoMS:MongoInstance 2021-03-18T18:54:55.414+0900 I CONTROL  [initandlisten] modules: none
  MongoMS:MongoInstance 2021-03-18T18:54:55.414+0900 I CONTROL  [initandlisten] build environment:"" +8ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:55.414+0900 I CONTROL  [initandlisten]     distarch: x86_64
  MongoMS:MongoInstance 2021-03-18T18:54:55.414+0900 I CONTROL  [initandlisten]     target_arch: x86_64
  MongoMS:MongoInstance 2021-03-18T18:54:55.414+0900 I CONTROL  [initandlisten] options: { net: { bindIp: "127.0.0.1", port: 56498 }, replication: { replSet: "testset" }, security: { authorization: "disabled" }, storage: { dbPath: "/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--59874-ejK88nw0MeZP", engine: "wiredTiger" } }"" +1ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:55.415+0900 I STORAGE  [initandlisten] wiredtiger_open config: create,cache_size=15872M,cache_overflow=(file_max=0M),session_max=20000,eviction=(threads_min=4,threads_max=4),config_base=false,statistics=(fast),log=(enabled=true,archive=true,path=journal,compressor=snappy),file_manager=(close_idle_time=100000),statistics_log=(wait=0),verbose=(recovery_progress),"" +0ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.005+0900 I STORAGE  [initandlisten] WiredTiger message [1616061296:5707][59899:0x11b048e00], txn-recover: Set global recovery timestamp: 0"" +591ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.175+0900 I RECOVERY [initandlisten] WiredTiger recoveryTimestamp. Ts: Timestamp(0, 0)"" +170ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.324+0900 I SHARDING [initandlisten] Marking collection local.system.replset as collection version: <unsharded>"" +148ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.324+0900 I SHARDING [initandlisten] Marking collection admin.system.roles as collection version: <unsharded>"" +0ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.324+0900 I SHARDING [initandlisten] Marking collection admin.system.version as collection version: <unsharded>"" +1ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.325+0900 I STORAGE  [initandlisten] createCollection: local.startup_log with generated UUID: 023e3702-e0b0-4d96-b5ab-daa971e0a08b"" +0ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.458+0900 I SHARDING [initandlisten] Marking collection local.startup_log as collection version: <unsharded>"" +133ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.458+0900 I FTDC     [initandlisten] Initializing full-time diagnostic data capture with directory '/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--59874-ejK88nw0MeZP/diagnostic.data'"" +0ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.459+0900 I STORAGE  [initandlisten] createCollection: local.replset.oplogTruncateAfterPoint with generated UUID: 6fa1509d-e483-44b5-b868-5687cbb529f1"" +1ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.554+0900 I STORAGE  [initandlisten] createCollection: local.replset.minvalid with generated UUID: f53c6362-1180-4e43-a149-28abe3040d73"" +96ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.677+0900 I SHARDING [initandlisten] Marking collection local.replset.minvalid as collection version: <unsharded>"" +122ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.677+0900 I STORAGE  [initandlisten] createCollection: local.replset.election with generated UUID: 560be7a5-e650-4789-867d-4c00286e4e85"" +0ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.775+0900 I SHARDING [initandlisten] Marking collection local.replset.election as collection version: <unsharded>"" +99ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.776+0900 I REPL     [initandlisten] Did not find local initialized voted for document at startup."" +0ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.776+0900 I REPL     [initandlisten] Did not find local Rollback ID document at startup. Creating one.
  MongoMS:MongoInstance 2021-03-18T18:54:56.776+0900 I STORAGE  [initandlisten] createCollection: local.system.rollback.id with generated UUID: 759cab49-dd26-41f0-82ed-c74979349005"" +0ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.865+0900 I SHARDING [initandlisten] Marking collection local.system.rollback.id as collection version: <unsharded>
  MongoMS:MongoInstance 2021-03-18T18:54:56.865+0900 I REPL     [initandlisten] Initialized the rollback ID to 1"" +89ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.865+0900 I REPL     [initandlisten] Did not find local replica set configuration document at startup;  NoMatchingDocument: Did not find replica set configuration document in local.system.replset"" +0ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.865+0900 I CONTROL  [LogicalSessionCacheRefresh] Sessions collection is not set up; waiting until next sessions refresh interval: Replication has not yet been configured"" +0ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.865+0900 I SHARDING [LogicalSessionCacheReap] Marking collection config.system.sessions as collection version: <unsharded>
  MongoMS:MongoInstance 2021-03-18T18:54:56.865+0900 I NETWORK  [initandlisten] waiting for connections on port 56498"" +0ms
  MongoMS:MongoInstance Mongo[56498]: Instance is ready! +2ms
  MongoMS:MongoInstance Mongo[56498]: run: Processes Started +0ms
  MongoMS:MongoMemoryServer _startUpInstance: Instance Started +6s
  MongoMS:MongoMemoryServer start: Instance fully Started +0ms
  MongoMS:MongoMemoryReplSet _initReplSet +6s
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.866+0900 I CONTROL  [LogicalSessionCacheReap] Sessions collection is not set up; waiting until next sessions reap interval: config.system.sessions does not exist"" +10ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.878+0900 I NETWORK  [listener] connection accepted from 127.0.0.1:56502 #1 (1 connection now open)"" +5ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:56.882+0900 I NETWORK  [conn1] received client metadata from 127.0.0.1:56502 conn1: { driver: { name: "nodejs", version: "3.6.3" }, os: { type: "Darwin", name: "darwin", architecture: "x64", version: "20.2.0" }, platform: "'Node.js v14.15.1, LE (unified)" }"" +0ms
  MongoMS:MongoInstance Mongo[56498]: STDOUT: ""2021-03-18T18:54:57.002+0900 I SHARDING [ftdc] Marking collection local.oplog.rs as collection version: <unsharded>"" +120ms
^[[A^[[A^[[A^[[B^[[BMongoServerSelectionError: Server selection timed out after 30000 ms
    at Timeout._onTimeout (/Users/sihoonkim/Documents/Rocketship/backend/node_modules/mongodb/lib/core/sdam/topology.js:438:30)
    at listOnTimeout (internal/timers.js:554:17)
    at processTimers (internal/timers.js:497:7) {
  reason: TopologyDescription {
    type: 'Unknown',
    setName: null,
    maxSetVersion: null,
    maxElectionId: null,
    servers: Map(1) { '127.0.0.1:56498' => [ServerDescription] },
    stale: false,
    compatible: true,
    compatibilityError: null,
    logicalSessionTimeoutMinutes: null,
    heartbeatFrequencyMS: 10000,
    localThresholdMS: 15,
    commonWireVersion: 7
  }
}
```

**Comment by hasezoey** [comment]
with this new debug log, i cannot see why there is an problem, from what i can tell, from where the timeout start counting, not even 30s have passed (in terms of what the log is outputting), the instance is started, and the replset should get initialized
-> i will look into this when i have the time, could you also test if it makes an difference if you use the latest beta?

**Comment by hoffnung8493** [comment]
ok ill try, and let you know. Thanks so far, for fast response :)

**Comment by hoffnung8493** [comment]
This is the debug log with the latest beta version(7.0.0-beta.33)

```
beforeAll
  MongoMS:MongoMemoryReplSet start +0ms
  MongoMS:MongoMemoryReplSet initAllServers +2ms
  MongoMS:MongoMemoryReplSet   starting server 1 of 1 +0ms
  MongoMS:MongoMemoryReplSet    instance opts: {
  auth: false,
  args: [],
  dbName: '08bf9dd4-993b-4cd9-a91a-2753de936e07',
  ip: '127.0.0.1',
  replSet: 'testset',
  storageEngine: 'wiredTiger'
} +0ms
  MongoMS:MongoMemoryReplSet initAllServers: waiting for all servers to finish starting +1ms
  MongoMS:MongoMemoryServer start: Called .start() method +0ms
  MongoMS:MongoMemoryServer _startUpInstance: Called MongoMemoryServer._startUpInstance() method +0ms
  MongoMS:MongoMemoryServer getStartOptions +0ms
  MongoMS:MongoMemoryServer _startUpInstance: Creating new MongoDB instance with options: {"instance":{"dbPath":"/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--61364-reSi9PM5RUTu","ip":"127.0.0.1","port":56596,"storageEngine":"wiredTiger","replSet":"testset","args":[],"auth":false},"binary":{},"spawn":{}} +4ms
  MongoMS:MongoInstance Mongo[56596]: run +0ms
  MongoMS:MongoBinary getPath +0ms
  MongoMS:DryMongoBinary generateDownloadPath: Generating Download Path, preferGlobal: "true" +0ms
  MongoMS:DryMongoBinary generateDownloadPath: Paths: {
  legacyHomeCache: '',
  modulesCache: '/Users/sihoonkim/Documents/Rocketship/backend/node_modules/.cache/mongodb-memory-server/mongod-x64-darwin-4.0.20',
  relative: '/Users/sihoonkim/Documents/Rocketship/backend/mongodb-binaries/mongod-x64-darwin-4.0.20',
  resolveConfig: ''
} +1ms
  MongoMS:DryMongoBinary generateDownloadPath: no existing binary for version "4.0.20" was found +0ms
  MongoMS:DryMongoBinary generateDownloadPath: using global (preferGlobal) "" +0ms
  MongoMS:MongoBinary getPath: MongoBinary options: {
  "version": "4.0.20",
  "downloadDir": ".",
  "os": {
    "os": "darwin"
  },
  "arch": "x64",
  "systemBinary": "",
  "platform": "darwin",
  "checkMD5": false
} +2ms
  MongoMS:DryMongoBinary locateBinary: Trying to locate Binary for version "4.0.20" +1ms
  MongoMS:DryMongoBinary generateDownloadPath: Generating Download Path, preferGlobal: "true" +0ms
  MongoMS:DryMongoBinary generatePaths: resolveConfigValue is not empty +0ms
  MongoMS:DryMongoBinary generateDownloadPath: Paths: {
  legacyHomeCache: '',
  modulesCache: '/Users/sihoonkim/Documents/Rocketship/backend/node_modules/.cache/mongodb-memory-server/mongod-x64-darwin-4.0.20',
  relative: '/Users/sihoonkim/Documents/Rocketship/backend/mongodb-binaries/mongod-x64-darwin-4.0.20',
  resolveConfig: '/Users/sihoonkim/Documents/Rocketship/backend/mongod-x64-darwin-4.0.20'
} +0ms
  MongoMS:DryMongoBinary generateDownloadPath: no existing binary for version "4.0.20" was found +0ms
  MongoMS:DryMongoBinary generateDownloadPath: using resolveConfig (DOWNLOAD_DIR) "/Users/sihoonkim/Documents/Rocketship/backend/mongod-x64-darwin-4.0.20" +1ms
  MongoMS:DryMongoBinary locateBinary: running generateDownloadPath +0ms
  MongoMS:DryMongoBinary generateDownloadPath: Generating Download Path, preferGlobal: "true" +0ms
  MongoMS:DryMongoBinary generatePaths: resolveConfigValue is not empty +0ms
  MongoMS:DryMongoBinary generateDownloadPath: Paths: {
  legacyHomeCache: '',
  modulesCache: '/Users/sihoonkim/Documents/Rocketship/backend/node_modules/.cache/mongodb-memory-server/mongod-x64-darwin-4.0.20',
  relative: '/Users/sihoonkim/Documents/Rocketship/backend/mongodb-binaries/mongod-x64-darwin-4.0.20',
  resolveConfig: '/Users/sihoonkim/Documents/Rocketship/backend/mongod-x64-darwin-4.0.20'
} +0ms
  MongoMS:DryMongoBinary generateDownloadPath: no existing binary for version "4.0.20" was found +0ms
  MongoMS:DryMongoBinary generateDownloadPath: using resolveConfig (DOWNLOAD_DIR) "/Users/sihoonkim/Documents/Rocketship/backend/mongod-x64-darwin-4.0.20" +0ms
  MongoMS:DryMongoBinary locateBinary: could not find an existing binary +0ms
  MongoMS:MongoBinary getPath: "RUNTIME_DOWNLOAD" is "true", trying to download +2ms
  MongoMS:MongoBinary download +0ms
  MongoMS:MongoBinary download: Waiting to acquire Download lock for file "/Users/sihoonkim/Documents/Rocketship/backend/4.0.20.lock" +1ms
  MongoMS:LockFile lock: Locking file "/Users/sihoonkim/Documents/Rocketship/backend/4.0.20.lock" +0ms
  MongoMS:LockFile checkLock: for file "/Users/sihoonkim/Documents/Rocketship/backend/4.0.20.lock" with uuid: "undefined" +0ms
  MongoMS:LockFile createLock: trying to create a lock file for "/Users/sihoonkim/Documents/Rocketship/backend/4.0.20.lock" +1ms
  MongoMS:LockFile createLock: Lock File Created for file "/Users/sihoonkim/Documents/Rocketship/backend/4.0.20.lock" +1ms
  MongoMS:MongoBinary download: Download lock acquired +2ms
  MongoMS:MongoBinary download: Adding version 4.0.20 to cache +0ms
  MongoMS:MongoBinaryDownload getMongodPath +0ms
  MongoMS:DryMongoBinary generateDownloadPath: Generating Download Path, preferGlobal: "true" +4ms
  MongoMS:DryMongoBinary generateDownloadPath: Paths: {
  legacyHomeCache: '',
  modulesCache: '/Users/sihoonkim/Documents/Rocketship/backend/node_modules/.cache/mongodb-memory-server/mongod-x64-darwin-4.0.20',
  relative: '/Users/sihoonkim/Documents/Rocketship/backend/mongodb-binaries/mongod-x64-darwin-4.0.20',
  resolveConfig: ''
} +0ms
  MongoMS:DryMongoBinary generateDownloadPath: no existing binary for version "4.0.20" was found +0ms
  MongoMS:DryMongoBinary generateDownloadPath: using global (preferGlobal) "" +0ms
  MongoMS:MongoBinaryDownload startDownload +2ms
  MongoMS:MongoBinaryDownloadUrl Using "mongodb-osx-ssl-x86_64-4.0.20.tgz" as the Archive String +0ms
  MongoMS:MongoBinaryDownloadUrl Using "https://fastdl.mongodb.org" as the mirror +0ms
  MongoMS:MongoBinaryDownload download +1ms
  MongoMS:MongoBinaryDownload download: Downloading: "https://fastdl.mongodb.org/osx/mongodb-osx-ssl-x86_64-4.0.20.tgz" +0ms
  MongoMS:MongoBinaryDownload httpDownload +1ms
  MongoMS:MongoBinaryDownload httpDownload: trying to download https://fastdl.mongodb.org/osx/mongodb-osx-ssl-x86_64-4.0.20.tgz +0ms
  MongoMS:MongoBinaryDownload httpDownload: moved "/Users/sihoonkim/Documents/Rocketship/backend/mongodb-osx-ssl-x86_64-4.0.20.tgz.downloading" to "/Users/sihoonkim/Documents/Rocketship/backend/mongodb-osx-ssl-x86_64-4.0.20.tgz" +2s
  MongoMS:MongoBinaryDownload makeMD5check: Checking MD5 of downloaded binary... +0ms
  MongoMS:MongoBinaryDownload makeMD5check: checkMD5 is disabled +0ms
  MongoMS:MongoBinaryDownload extract +0ms
  MongoMS:DryMongoBinary generateDownloadPath: Generating Download Path, preferGlobal: "true" +2s
  MongoMS:DryMongoBinary generateDownloadPath: Paths: {
  legacyHomeCache: '',
  modulesCache: '/Users/sihoonkim/Documents/Rocketship/backend/node_modules/.cache/mongodb-memory-server/mongod-x64-darwin-4.0.20',
  relative: '/Users/sihoonkim/Documents/Rocketship/backend/mongodb-binaries/mongod-x64-darwin-4.0.20',
  resolveConfig: ''
} +1ms
  MongoMS:DryMongoBinary generateDownloadPath: no existing binary for version "4.0.20" was found +0ms
  MongoMS:DryMongoBinary generateDownloadPath: using global (preferGlobal) "" +0ms
  MongoMS:MongoBinaryDownload extract: archive: "/Users/sihoonkim/Documents/Rocketship/backend/mongodb-osx-ssl-x86_64-4.0.20.tgz" final: "/Users/sihoonkim/Documents/Rocketship/backend/mongod-x64-darwin-4.0.20" +1ms
  MongoMS:MongoBinaryDownload extractTarGz +1ms
  MongoMS:MongoBinary download: Removing Download lock +3s
  MongoMS:LockFile unlock: Unlocking file "/Users/sihoonkim/Documents/Rocketship/backend/4.0.20.lock" +3s
  MongoMS:LockFile checkLock: for file "/Users/sihoonkim/Documents/Rocketship/backend/4.0.20.lock" with uuid: "0277e3e2-0af1-44f9-8094-f7b4517123b8" +0ms
  MongoMS:LockFile checkLock: Lock File Already exists, and is for *this* process, with uuid: "0277e3e2-0af1-44f9-8094-f7b4517123b8" +0ms
  MongoMS:LockFile unlock: Lock Status was "availableInstance" for file "/Users/sihoonkim/Documents/Rocketship/backend/4.0.20.lock" +0ms
  MongoMS:LockFile unlockCleanup: for file "/Users/sihoonkim/Documents/Rocketship/backend/4.0.20.lock" +1ms
  MongoMS:MongoBinary download: Download lock removed +2ms
  MongoMS:MongoBinary getPath: Mongod binary path: "/Users/sihoonkim/Documents/Rocketship/backend/mongod-x64-darwin-4.0.20" +0ms
  MongoMS:MongoInstance Mongo[56596]: run: Starting Processes +3s
  MongoMS:MongoInstance Mongo[56596]: _launchMongod: Launching Mongod Process +0ms
  MongoMS:MongoInstance Mongo[56596]: prepareCommandArgs +0ms
  MongoMS:MongoInstance Mongo[56596]: prepareCommandArgs: final arugment array:["--port","56596","--dbpath","/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--61364-reSi9PM5RUTu","--storageEngine","wiredTiger","--bind_ip","127.0.0.1","--noauth","--replSet","testset"] +0ms
  MongoMS:MongoInstance Mongo[56596]: _launchKiller: Launching Killer Process (parent: 61364, child: 61387) +11ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:03.114+0900 I CONTROL  [main] Automatically disabling TLS 1.0, to force-enable TLS 1.0 specify --sslDisabledProtocols 'none'"" +576ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:03.124+0900 I CONTROL  [initandlisten] MongoDB starting : pid=61387 port=56596 dbpath=/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--61364-reSi9PM5RUTu 64-bit host=gimsihun-ui-MacBookPro.local
  MongoMS:MongoInstance 2021-03-18T19:09:03.124+0900 I CONTROL  [initandlisten] db version v4.0.20
  MongoMS:MongoInstance 2021-03-18T19:09:03.124+0900 I CONTROL  [initandlisten] git version: e2416422da84a0b63cde2397d60b521758b56d1b
  MongoMS:MongoInstance 2021-03-18T19:09:03.124+0900 I CONTROL  [initandlisten] allocator: system"" +9ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:03.124+0900 I CONTROL  [initandlisten] modules: none
  MongoMS:MongoInstance 2021-03-18T19:09:03.124+0900 I CONTROL  [initandlisten] build environment:
  MongoMS:MongoInstance 2021-03-18T19:09:03.124+0900 I CONTROL  [initandlisten]     distarch: x86_64
  MongoMS:MongoInstance 2021-03-18T19:09:03.124+0900 I CONTROL  [initandlisten]     target_arch: x86_64
  MongoMS:MongoInstance 2021-03-18T19:09:03.124+0900 I CONTROL  [initandlisten] options: { net: { bindIp: "127.0.0.1", port: 56596 }, replication: { replSet: "testset" }, security: { authorization: "disabled" }, storage: { dbPath: "/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--61364-reSi9PM5RUTu", engine: "wiredTiger" } }"" +0ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:03.125+0900 I STORAGE  [initandlisten] wiredtiger_open config: create,cache_size=15872M,cache_overflow=(file_max=0M),session_max=20000,eviction=(threads_min=4,threads_max=4),config_base=false,statistics=(fast),log=(enabled=true,archive=true,path=journal,compressor=snappy),file_manager=(close_idle_time=100000),statistics_log=(wait=0),verbose=(recovery_progress),"" +1ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:03.680+0900 I STORAGE  [initandlisten] WiredTiger message [1616062143:680646][61387:0x117319e00], txn-recover: Set global recovery timestamp: 0"" +555ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:03.853+0900 I RECOVERY [initandlisten] WiredTiger recoveryTimestamp. Ts: Timestamp(0, 0)"" +173ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:03.993+0900 I SHARDING [initandlisten] Marking collection local.system.replset as collection version: <unsharded>"" +140ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:03.993+0900 I SHARDING [initandlisten] Marking collection admin.system.roles as collection version: <unsharded>"" +0ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:03.994+0900 I SHARDING [initandlisten] Marking collection admin.system.version as collection version: <unsharded>"" +1ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:03.994+0900 I STORAGE  [initandlisten] createCollection: local.startup_log with generated UUID: 25b38cf8-fb63-4e6d-80c8-201e94298afe"" +0ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:04.114+0900 I SHARDING [initandlisten] Marking collection local.startup_log as collection version: <unsharded>"" +120ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:04.114+0900 I FTDC     [initandlisten] Initializing full-time diagnostic data capture with directory '/var/folders/rq/b244xr2135x5mwd9dtn80flc0000gn/T/mongo-mem--61364-reSi9PM5RUTu/diagnostic.data'"" +0ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:04.115+0900 I STORAGE  [initandlisten] createCollection: local.replset.oplogTruncateAfterPoint with generated UUID: 39b0e449-fd15-44fa-a269-4ee9bf790493"" +1ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:04.213+0900 I STORAGE  [initandlisten] createCollection: local.replset.minvalid with generated UUID: 8fca9250-06ef-440b-86ae-38cecfddfcc8"" +98ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:04.338+0900 I SHARDING [initandlisten] Marking collection local.replset.minvalid as collection version: <unsharded>"" +125ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:04.338+0900 I STORAGE  [initandlisten] createCollection: local.replset.election with generated UUID: 902093e4-8922-4494-9f54-62e9c046b6df"" +0ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:04.469+0900 I SHARDING [initandlisten] Marking collection local.replset.election as collection version: <unsharded>"" +131ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:04.469+0900 I REPL     [initandlisten] Did not find local initialized voted for document at startup.
  MongoMS:MongoInstance 2021-03-18T19:09:04.469+0900 I REPL     [initandlisten] Did not find local Rollback ID document at startup. Creating one.
  MongoMS:MongoInstance 2021-03-18T19:09:04.469+0900 I STORAGE  [initandlisten] createCollection: local.system.rollback.id with generated UUID: 850b48c9-e099-48d9-8ed2-ec8756d7638b"" +1ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:04.577+0900 I SHARDING [initandlisten] Marking collection local.system.rollback.id as collection version: <unsharded>"" +107ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:04.577+0900 I REPL     [initandlisten] Initialized the rollback ID to 1
  MongoMS:MongoInstance 2021-03-18T19:09:04.577+0900 I REPL     [initandlisten] Did not find local replica set configuration document at startup;  NoMatchingDocument: Did not find replica set configuration document in local.system.replset"" +1ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:04.578+0900 I CONTROL  [LogicalSessionCacheRefresh] Sessions collection is not set up; waiting until next sessions refresh interval: Replication has not yet been configured"" +0ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:04.578+0900 I SHARDING [LogicalSessionCacheReap] Marking collection config.system.sessions as collection version: <unsharded>
  MongoMS:MongoInstance 2021-03-18T19:09:04.578+0900 I NETWORK  [initandlisten] waiting for connections on port 56596"" +0ms
  MongoMS:MongoInstance Mongo[56596]: Instance is ready! +0ms
  MongoMS:MongoInstance Mongo[56596]: run: Processes Started +0ms
  MongoMS:MongoMemoryServer _startUpInstance: Instance Started +5s
  MongoMS:MongoMemoryServer start: Instance fully Started +0ms
  MongoMS:MongoMemoryReplSet _initReplSet +5s
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:04.578+0900 I CONTROL  [LogicalSessionCacheReap] Sessions collection is not set up; waiting until next sessions reap interval: config.system.sessions does not exist"" +10ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:04.588+0900 I NETWORK  [listener] connection accepted from 127.0.0.1:56600 #1 (1 connection now open)
  MongoMS:MongoInstance 2021-03-18T19:09:04.593+0900 I NETWORK  [conn1] received client metadata from 127.0.0.1:56600 conn1: { driver: { name: "nodejs", version: "3.6.3" }, os: { type: "Darwin", name: "darwin", architecture: "x64", version: "20.2.0" }, platform: "'Node.js v14.15.1, LE (unified)" }"" +5ms
  MongoMS:MongoInstance Mongo[56596]: STDOUT: ""2021-03-18T19:09:05.006+0900 I SHARDING [ftdc] Marking collection local.oplog.rs as collection version: <unsharded>"" +413ms
MongoServerSelectionError: Server selection timed out after 30000 ms
    at Timeout._onTimeout (/Users/sihoonkim/Documents/Rocketship/backend/node_modules/mongodb/lib/core/sdam/topology.js:438:30)
    at listOnTimeout (internal/timers.js:554:17)
    at processTimers (internal/timers.js:497:7) {
  reason: TopologyDescription {
    type: 'Unknown',
    setName: null,
    maxSetVersion: null,
    maxElectionId: null,
    servers: Map(1) { '127.0.0.1:56596' => [ServerDescription] },
    stale: false,
    compatible: true,
    compatibilityError: null,
    logicalSessionTimeoutMinutes: null,
    heartbeatFrequencyMS: 10000,
    localThresholdMS: 15,
    commonWireVersion: 7
  }
}
```

**Comment by hasezoey** [comment]
sorry, i take back my statement this "30s" timeout is not related to the fix, this is the `MongoClient.connect` failing, not `_waitForPrimary` timeout

i could not find any reason on why it times out, could you try out `7.0.0-beta.34` and fully-reinstall all node_modules & report back? that release includes more logs and upgrades the mongodb dependencie

**Comment by jeffontheground** [comment]
I ran into the same issue (Server selection timed out after 30000 ms) on macOS High Sierra with 7.0.0-beta.26, node 14.15.3
After wasting 4 hours, I finally work it out.
Just
1. remove  node_modules
2. remove yarn.lock
3. reinstall node_modules
Done
Hope this will help you

**Comment by simllll** [comment]
Can we reopen this? I'm experience the same from time to time.. which  means our tests are super flacky right now.
MongoBulkWriteError: Server selection timed out after 30000 ms

only happens with replset though.

**Comment by hasezoey** [comment]
@simllll please open a new issue, this issue is already quite old (and for reference, reference this old issue)

also include all required information from the template, with [debug logs](https://nodkz.github.io/mongodb-memory-server/docs/guides/enable-debug-mode) (not trimmed)

**Comment by simllll** [comment]
Follow up here https://github.com/nodkz/mongodb-memory-server/issues/733
